### PR TITLE
Allows partial assembly with nonconstant scalar coefficients for Gradient and VectorDiffusion integrators [pa-coeffs-nonconstant]

### PR DESCRIPTION
**Pull request by smattis-nnl**
Currently `GradientIntegrator` and `VectorDiffusionIntegrator` only allow for constant scalar coefficients for partial assembly.  This PR adds the ability to use nonconstant scalar coefficients for partial assembly with these integrators. It does this in a similar way as is implemented for `DiffusionIntegrator` in `bilininteg_diffusion_pa.cpp`. A vector is used to store the coefficient values. The vectors are reshaped into 2d arrays. For constant coefficients, it is a 1x1 array. For nonconstant coefficients, it uses an nq x ne array where nq is the number of quadrature points and ne is the number of elements. It includes checks to verify that the coefficient input can be used.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2270](https://github.com/mfem/mfem/pull/2270) | @smattis-nnl | @tzanio | @jandrej + @pazner | 05/27/21 | 06/11/21 | 06/14/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by jandrej — APPROVED** [review]

File: fem/bilininteg_vecdiffusion.cpp [hunk 1]
```diff
@@ -166,6 +216,9 @@ void VectorDiffusionIntegrator::AssemblePA(const FiniteElementSpace &fes)
       auto W = w.Read();
       auto J = Reshape(j.Read(), NQ, SDIM, DIM, ne);
       auto D = Reshape(d.Write(), NQ, SDIM, ne);
+
+      auto C = Reshape(coeff.HostWrite(), NQ, ne);
```
Comment (pazner) [inline fem/bilininteg_vecdiffusion.cpp:220]: Shouldn't this be `coeff.Read()` and doesn't it need to take into account the special case where `coeff` has only one entry?

Comment (smattis-nnl) [reply fem/bilininteg_vecdiffusion.cpp]: @pazner Yes. Thanks for catching that. This is fixed by a250079153d197c1ee6cfb1e3b7540e7dd839668.

**Review by pazner — APPROVED** [review]
This is a nice addition. Thanks for the contribution @smattis-nnl!

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Review by tzanio — APPROVED** [review]
Thanks for this contribution @smattis-nnl !

@jandrej and @pazner -- thank you for reviewing.

**Pull request merged**